include/groov/mmio_bus.hpp

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include <async/concepts.hpp>
+#include <async/just.hpp>
+#include <async/then.hpp>
+
+#include <bit>
+#include <cstdint>
+
+namespace groov {
+namespace detail {
+template <std::size_t NumBits> struct uint_of_width;
+
+template <> struct uint_of_width<8> {
+    using type = std::uint8_t;
+};
+
+template <> struct uint_of_width<16> {
+    using type = std::uint16_t;
+};
+
+template <> struct uint_of_width<32> {
+    using type = std::uint32_t;
+};
+
+template <> struct uint_of_width<64> {
+    using type = std::uint64_t;
+};
+
+template <std::size_t NumBits>
+using uint_of_width_t = typename uint_of_width<NumBits>::type;
+} // namespace detail
+
+struct cpp_mem_iface {
+    template <std::unsigned_integral T> static auto store(std::uintptr_t addr) {
+        return async::then([=](T value) -> void {
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+            *reinterpret_cast<T volatile *>(addr) = value;
+        });
+    }
+
+    template <std::unsigned_integral T>
+    static auto load(std::uintptr_t addr) -> async::sender auto {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
```

Comment:
Question: Doesn't `load` need to return a lazy sender like `store`. For example, wrapping in `async::then`?

I might be thinking about this wrong, but wouldn't,

```c++
   auto s = read(0x8000'1000) | async::then(do_stuff);
```

evaluate and read 0x8000'0000 at construction and then populate a `just` sender that will have the value from construction and not at the point where the sender is invoked?

Reply:
good catch!